proposals/self-constraint.md

```diff
@@ -0,0 +1,123 @@
+
+The following syntax would become valid:
+```csharp
+interface IAdditionOperators<this TSelf, TOther, TResult>
+{
+    static abstract TResult operator +(TSelf left, TOther right);
+}
+
+interface IAdditiveIdentity<this TSelf>
+{
+    static abstract TSelf AdditiveIdentity { get; }
+}
+```
+
+This would cause the constraint for the `TSelf` generic type parameter to have a `modreq` emitted (similarly to how the `modreq` for the `unmanaged` constraint works) that indicates it is the "self type" as well as the standard recursive constraint that was already being declared beforehand. The language can then depend on this information when interacting with the interface and know, for example, that `IAdditionOperators<int, TOther, TResult>` should resolve to the corresponding APIs exposed on `int`.
+
+Only a single generic-type parameter would be allowed to be annotated as the `self type` as there is no reason for two types representing the same contract.
+
+The self-constrained type argument must still be specified at the use site. That is the user must still fully specify `struct Half : IAdditiveIdentity<Half>` and could not choose to elide or otherwise not specify `<Half>`.
+
+** TODO: Insert list of actual spec changes required here **
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+The new syntax may be confusing to some users as the constraint is now specified in a different location.
+
+Utilizing the new syntax would be a breaking change for existing types and so you could not, for example, modify `System.IEquatable<T>` to be `System.IEquatable<this T>`.
+
+## Alternatives
+[alternatives]: #alternatives
+
+Consider exposing the constraint in a different manner such as:
+```csharp
+interface IAdditiveIdentity<TSelf>
+    where TSelf : this
+{
+    static abstract TSelf AdditiveIdentity { get; }
+}
+```
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+Should you be able to place additional constraints on the `self` type? For example, would `this TSelf` and `where TSelf : this, struct` be valid if you only wanted the interface to be implementable by struct types?
+
+Does this need to be a `modopt` (or `modreq`?) or can it be expressed in some way such that existing types can "move forward" if they have existing scenarios where this was the intended contract?
+
+How does this play into variance for classes and possible eventual support for static abstracts in classes? Would variance conversions between interface types lead to a loss of information in some way or another? Etc.
+
+Would a proper self constraint allow a simpler model for encoding the `call` in IL? That is, if the runtime also supports a proper self constraint then rather than requiring ``constrained. !!T call !0 class IAdditiveIdentity`1<!!T>.get_AdditiveIdentity()`` it could simply support ``call !0 class IAdditiveIdentity`1<!!T>.get_AdditiveIdentity()``.
+
+Could the self type be represented as an associate type rather than as a type parameter to an interface?
```

Comment:
```suggestion
Could the self type be represented as an associate type rather than as a type parameter to an interface?

Obligatory general syntax questions: should we use a keyword other than `this`? Should the keyword go where the proposal puts it, or with the rest of the type constraints?
```

Reply:
Is the second part necessary in relation to the existing `## Alternatives` section showing having it with the rest of the type constraints?

Reply:
I did not interpret the existing question as a matter of syntax: I interpreted it as a matter of semantics. This is specifically a question about syntax.